### ISSUE TITLE
Update Mozilla/MS stores

### ISSUE THREAD

**Issue by chrcoluk** [drwetter/testssl.sh#317]
Using a letsencrypt cert with the root and intermediate cert added to the cert file, I get the following result on the certificate test.

 # of certificates provided   3
 Chain of trust (experim.)    NOT ok: microsoft (chain incomplete)
                                OK: mozilla linux

Yet on microsoft outlook there is no certificate warning, suggesting the certificate chain is fine on microsoft.

Regards

Chris

**Comment by Niko78** [comment]
I have the same issue with  a cert issuer : "TERENA SSL CA 2" ("TERENA" from "NL")

I get this result : 
NOT ok: Microsoft (chain incomplete) Mozilla (chain incomplete)
OK: Linux

no issue at all on microsoft and mozilla
maybe it would be interesting to update pem files ?

Nicolas

**Comment by drwetter** [comment]
> # of certificates provided 3
. Chain of trust (experim.) NOT ok: microsoft (chain incomplete)
> OK: mozilla linux
> 
> Yet on microsoft outlook there is no certificate warning, suggesting the certificate chain is fine on > microsoft.

Each browser pulls necessary intermediate certs. So a test by just using a browser.whether the certificate chain is complete is unfortunately wrong.

The MS chain was updated a bit longer than a month ago. Can't tell however whether both possible paths of letsencrypt were in there.

**Comment by drwetter** [comment]
@Niko78 :  I need an IP address .

**Comment by chrcoluk** [comment]
This isnt a browser but a email client.

try mail.freebsd-admin.com

**Comment by chrcoluk** [comment]
thanks, if you like I can retest on latest code?

**Comment by drwetter** [comment]
Am 13. März 2016 14:17:05 MEZ, schrieb chrcoluk <notifications@github.com>:
>thanks, if you like I can retest on latest code?

Chris, go ahead.

@niko78 The host you gave me still has a problem with the MS store. MS decided after XP to update the root and 3rd party store on demand and not via update -- which is broken by design.  Here I need to find a good solution for testssl.sh.

Cheers, Dirk 


-- 
Sent from my mobile

**Comment by Niko78** [comment]
Dirk,

with this file : https://raw.githubusercontent.com/markofu/scripts/master/nmap/nmap/ncat/certs/ca-bundle.crt
=> it works

but yesterday I extract all CA cert from my win10 up to date and you are right it did not work with this extract

as it misses an intermediate CA to validate the certificate, testssl.sh should warn us... or I add this extra CA on my website to be sent with the issuer already sent

nico

**Comment by drwetter** [comment]
Am 14. März 2016 07:02:53 MEZ, schrieb Niko78 <notifications@github.com>:
>Dirk,
>
>with this file :
>https://raw.githubusercontent.com/markofu/scripts/master/nmap/nmap/ncat/certs/ca-bundle.crt
>=> it works

Maybe  :)  But I would rather rely not on 3rd parties here. Also for future updates it's not what I would do.

Probably I would need to add all Third Party CAs from certmgr.msc .

>but yesterday I extract all CA cert from my win10 up to date and you
>are right it did not work with this extract
>
>as it misses an intermediate CA to validate the certificate, testssl.sh
>should warn us... or I add this extra CA on my website to be sent with
>the issuer already sent

It should say "chain not complete". The problem is not on your side, so I suggest to wait until it has been addressed at the source.

If somebody else spotted an issue with the Microsoft store: please speak up :-)

Cheers! Dirk  


-- 
Sent from my mobile

**Comment by chrcoluk** [comment]
thanks, looks good now. :)

 Signature Algorithm          SHA256 with RSA
 Server key size              3072 bits
 Fingerprint / Serial         SHA1 6644DCEFF5C86BA894A971B9417925E7AF595398 / 014C123DCE9297A54F0F6C1BA335F1B68E37
                              SHA256 0258777EE7447D192282424B3B91B5ED840FC03D89BF389F192DC557BF398E1A
 Common Name (CN)             "freebsd-admin.com" (works w/o SNI)
 subjectAltName (SAN)         "freebsd-admin.com" "mail.freebsd-admin.com" "www.freebsd-admin.com" 
 Issuer                       "Let's Encrypt Authority X1" ("Let's Encrypt" from "US")
 EV cert (experimental)       no 
 Certificate Expiration       84 >= 60 days (2016-03-09 11:11 --> 2016-06-07 12:11 +0100)
 # of certificates provided   3
 Chain of trust (experim.)    Ok   
 Certificate Revocation List  --
 OCSP URI                     http://ocsp.int-x1.letsencrypt.org/
 OCSP stapling                --

**Comment by drwetter** [comment]
can some body please help here with the Apple certificate store?

**Comment by Niko78** [comment]
see below, last pem from Sierra

[apple.zip](https://github.com/drwetter/testssl.sh/files/502320/apple.zip)

**Comment by drwetter** [comment]
Thank you.

Have you imported a CA or is that 1:1 from Apple? Just wondering what the 4th CA is about:

```
openssl x509 -in level4.crt -text -noout                                                                        
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 1005814224 (0x3bf381d0)
    Signature Algorithm: sha1WithRSAEncryption
        Issuer: C=ch, O=admin, OU=Services, OU=Certification Authorities, CN=Admin-Root-CA
        Validity
            Not Before: Nov 15 08:51:07 2001 GMT
            Not After : Nov 10 07:51:07 2021 GMT
        Subject: C=ch, O=admin, OU=Services, OU=Certification Authorities, CN=Admin-Root-CA
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:cb:e0:af:44:14:22:fe:6a:17:49:f2:5d:9d:cf:
                    5c:02:62:0b:80:b9:5a:bf:a8:55:33:7b:a3:12:24:
                    5e:4c:6f:3b:41:89:4e:83:3d:9f:b3:95:26:34:a6:
[..]
                  12:a8:8b:21:c6:42:ae:7a:64:af:f2:b3:c0:ed:77:
                    e4:69
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Certificate Policies:
                Policy: 2.16.756.1.17.3.1.0
                  User Notice:
                    Explicit Text: This is the Admin-Root-CA CPS
                  CPS: http://www.informatik.admin.ch/PKI/links/CPS[..]pdf
```

Can anybody else confirm the correctness of this root CA please?

**Comment by Niko78** [comment]
it is well in the mac os Sierra CA list
https://support.apple.com/en-us/HT207189

Admin-Root-CA 	Admin-Root-CA 	RSA 	2048 bits 	SHA-1 	3B F3 81 D0 	07:51:07 Nov 10, 2021 	Not EV

Nicolas

**Comment by drwetter** [comment]
(merci) 

WTF. The DN of CA looked like an apprentice issued a test CA for his purposes. As it turns out though It seems to belong to Swiss government and Apple doesn't seem to have a transparent policy  -- as opposed to Mozilla: https://bugzilla.mozilla.org/show_bug.cgi?id=639763 .

**Comment by drwetter** [comment]
updated, see https://github.com/drwetter/testssl.sh/commit/eb1f6e05bb8440ab8cb4e70b3e0308ee5c7aabb3.

Open: MS

**Comment by drwetter** [comment]
update MS store, see d32dbda .

Not really happy with it as the root certificates which MS lists @ http://social.technet.microsoft.com/wiki/contents/articles/31634.microsoft-trusted-root-certificate-program-participants-v-2016-april.aspx are more than I pushed.

If anybody has a clue how to do this properly (what I did, see https://github.com/drwetter/testssl.sh/blob/master/etc/README.md) I am glad to learn.

Supposedly there's somewhere a download but I wasn't able to find it.
